qcodes/dataset/sqlite/queries.py

```diff
@@ -461,27 +472,97 @@ def get_parameter_tree_values(
         end: The (1-indexed) result to include as the last result to be
             returned. None is equivalent to "all the rest". If start > end,
             nothing is returned.
+        callback: Function called during the data loading every
+            config.dataset.callback_percent.
 
     Returns:
         A list of list. The outer list index is row number, the inner list
         index is parameter value (first toplevel_param, then other_param_names)
     """
 
+    cursor = conn.cursor()
+
+    offset: int | np.ndarray
+
     offset = max((start - 1), 0) if start is not None else 0
     limit = max((end - offset), 0) if end is not None else -1
 
     if start is not None and end is not None and start > end:
         limit = 0
 
+    # start and end currently not working with callback
+    if start is None and end is None and callback is not None:
+
```

Comment:
Looking at this again I am a bit confused about the calculations. Max(ID) will return the number of rows in the table and each row is fetched as one entry so why is the need to incorporate the number of dependent parameters in the calculation?

Reply:
I think it would also make the code somewhat clear if we break this logic to calculate offset out into one or more functions

Reply:
My understanding is that when a user downloads a certain parameter, a lot of rows are useless to him since containing `NULL` values:
![image](https://user-images.githubusercontent.com/13218882/201927074-8866e591-2b8b-48ba-903f-a721154273e0.png)
Thus the number of data points is the total number of row divided by the number of dependent parameters, the  number of column.
That ratio `int(max_id/nb_param_dependent)` gives then the true number of row being downloaded.

The main idea here is that we are divided one big SQL request into many smaller SQL requests.
By doing so, we need to keep track of how many request are being done and where we are in that process.
By using `LIMIT` and `OFFSET`, we can do so but need the true number of downloaded rows.

Reply:
That sounds correct if each parameter has been measured the same number of times but need not be case. Ideally we want to know the number of rows where dependent_parameter is not Null. One could perhaps use count to get the number of rows where dependent_parameter is not Null. See https://www.w3schools.com/sql/sql_count_avg_sum.asp

Reply:
Yes, I think you're right.
I have break down the logic into smaller functions and change the way I count the number of row to be downloaded.